src/soilId/soilIdSlice.ts

```diff
@@ -158,32 +163,67 @@ const soilIdSlice = createSlice({
       state.status = 'ready';
     });
 
-    builder.addCase(fetchSoilIdMatches.pending, (state, action) => {
-      state.soilIdParams = {
-        coords: action.meta.arg.coords,
-        siteId: action.meta.arg.siteId,
-      };
-      state.soilIdData = initialState.soilIdData;
-      state.soilIdStatus = 'loading';
+    builder.addCase(fetchLocationBasedSoilMatches.pending, (state, action) => {
+      const key = soilIdKey(action.meta.arg);
+      state.matches[key] = soilIdEntryForStatus('loading');
+    });
+
+    builder.addCase(fetchLocationBasedSoilMatches.rejected, (state, action) => {
+      const key = soilIdKey(action.meta.arg);
+      state.matches[key] = soilIdEntryForStatus('error');
+    });
+
+    builder.addCase(
+      fetchLocationBasedSoilMatches.fulfilled,
+      (state, action) => {
+        const key = soilIdKey(action.meta.arg);
+        if (action.payload.__typename === 'SoilIdFailure') {
+          state.matches[key] = soilIdEntryForStatus(action.payload.reason);
+        } else {
+          state.matches[key] = soilIdEntryLocationBased(action.payload.matches);
+        }
+      },
+    );
+
+    builder.addCase(fetchDataBasedSoilMatches.pending, (state, action) => {
+      const key = soilIdKey(action.meta.arg.coords, action.meta.arg.siteId);
+      state.matches[key] = soilIdEntryForStatus('loading');
     });
 
-    builder.addCase(fetchSoilIdMatches.rejected, state => {
-      state.soilIdStatus = 'error';
+    builder.addCase(fetchDataBasedSoilMatches.rejected, (state, action) => {
+      const key = soilIdKey(action.meta.arg.coords, action.meta.arg.siteId);
+      state.matches[key] = soilIdEntryForStatus('error');
     });
 
-    builder.addCase(fetchSoilIdMatches.fulfilled, (state, action) => {
-      const { failureReason, ...payload } = action.payload;
-      state.soilIdStatus = failureReason ?? 'ready';
-      state.soilIdData = payload;
+    builder.addCase(fetchDataBasedSoilMatches.fulfilled, (state, action) => {
+      const key = soilIdKey(action.meta.arg.coords, action.meta.arg.siteId);
+      if (action.payload.__typename === 'SoilIdFailure') {
+        state.matches[key] = soilIdEntryForStatus(action.payload.reason);
+      } else {
+        state.matches[key] = soilIdEntryDataBased(action.payload.matches);
+      }
     });
   },
 });
 
+const flushDataBasedMatches = (state: SoilState) => {
+  /*
+   * When soil ID input data changes (e.g. samples, intervals), we clear any
+   * cached entries that are data-based since they aren't valid anymore.
+   */
+  Object.keys(state.matches)
+    .map(key => key as SoilIdKey)
```

Comment:
nit: i'm not sure if typescript/javascript will be smart enough to elide this at runtime. but there are probably bigger CPU cycle fishes to fry

Reply:
I can refactor this to remove that line.